File: sdk/reproit-backend-rb/lib/reproit_backend_rb/ci.rb
# CI capture mode for reproit-backend-rb: the flaky-CI wedge.
#
# `CI.suite(name)` returns a Minitest-backed `t.call(name) { ... }` whose
# trigger identity is the TEST (suite plus test id), not an inbound HTTP
# request. With `REPROIT_CI_CAPTURE=1` every test runs inside its own trace,
# so the instrumented Net::HTTP hook and db helper (instrument.rb) record
# dependency exchanges and the determinism envelope exactly as production
# capture does; a FAILING test emits a version-2 `reproit-backend-capture`
# capsule to a bounded on-disk spool. With `REPROIT_REPLAY` set the SAME
# wrapper re-runs only the capsule's named test while the SDK serves the
# recorded exchanges in process, and reports the observed result as a
# structured stderr marker for `reproit check`. Without either env the
# wrapper is plain Minitest untouched.
#
# The wire is the existing capture payload: the test identity rides in the
# `operation` field as `test:<suite>#<test>`, and the failed assertion is
# the existing `backend-authored-invariant` registry oracle (a test IS an
# authored invariant). No new protocol fields, no new oracle ids.
#
# Honest limit: replay pins the envelope and the recorded exchanges, which
# is the whole boundary this SDK can see. A race the boundary cannot see
# (scheduling, shared memory) is not reproduced by this capsule; `reproit
# check` reports such runs Inconclusive, never a fake reproduction.
#
# Loaded via `ReproitBackendRb.autoload`, never by requiring the SDK: this
# file pulls in minitest/autorun, which installs an at_exit test runner no
# production host should inherit.

require "digest"
require "fileutils"
require "json"
require "minitest/autorun"

require_relative "trace"
require_relative "capture"
require_relative "instrument"

module ReproitBackendRb
  module CI
    # Test-trigger identity prefix inside the existing `operation` field.
    TEST_TRIGGER_PREFIX = "test:"
    # The registry oracle a failed test capsule carries: an authored
    # invariant (the test's own assertion) was violated. Existing id, not a
    # new one.
    TEST_FAILURE_ORACLE = "backend-authored-invariant"
    # Structured stderr markers `reproit check` parses, like
    # REPROIT:DIVERGENCE.
    RESULT_MARKER = "REPROIT:CI-TEST "
    SPOOL_MARKER = "REPROIT:CI-CAPSULE "

    # Spool bounds. The cap covers the TOTAL bytes on disk; spilled capsules
    # beyond it are dropped and counted (in-process stats plus the on-disk
    # `dropped.count`), never silently.
    DEFAULT_SPOOL_DIR = ".reproit/ci-spool"
    DEFAULT_SPOOL_MAX_BYTES = 16 * 1024 * 1024
    SPOOL_MAX_FLOOR_BYTES = 4 * 1024
    SPOOL_MAX_CEIL_BYTES = 64 * 1024 * 1024
    # Suite and test names share the operation field's 256-code-point bound.
    MAX_NAME = 120
    MAX_ERROR_CHARS = 2048

    @lock = Mutex.new
    @trace_seq = 0
    @suite_seq = 0
    @stats = { spooled_capsules: 0, dropped_capsules: 0, failed_captures: 0 }
    @replay_target = nil
    @replay_target_loaded = false

    module_function

    def stats
      @lock.synchronize { @stats.dup }
    end

    def count(key)
      @lock.synchronize { @stats[key] += 1 }
    end

    def replay_path
      value = ENV["REPROIT_REPLAY"].to_s
      value.empty? ? nil : value
    end

    def mode
      return "replay" unless replay_path.nil?
      return "capture" if ENV["REPROIT_CI_CAPTURE"] == "1"
      "off"
    end

    def bounded_name(value)
      value.to_s.strip[0, MAX_NAME]
    end

    def operation_for(suite_name, test_name)
      TEST_TRIGGER_PREFIX + bounded_name(suite_name) + "#" + bounded_name(test_name)
    end

    def bounded_error(error)
      message = error.respond_to?(:message) ? error.message : error
      message.to_s[0, MAX_ERROR_CHARS]
    end

    # Synthesized trace context: the CI job stands where production stood.
    def ci_context
      seq = @lock.synchronize { @trace_seq += 1 }
      commit = [ENV["REPROIT_COMMIT"], ENV["GITHUB_SHA"]].find do |value|
        ReproitBackendRb.valid_token?(value)
      end
      {
        "trace_id" => format("ci-%d-%d", (Time.now.to_f * 1000).to_i, seq),
        "actor" => nil,
        "action_index" => 0,
        "build" => commit,
        "config_contract" => nil,
        "capture_envelope" => true,
      }
    end

    # Same envelope shape production capture records; the seed pins the
    # REPLAY run's randomness, it does not reproduce the test run's.
    def envelope_for(trace)
      first = trace.events[0] || {}
      ReproitBackendRb.determinism_envelope(first["at"].is_a?(Integer) ? first["at"] : nil)
    end

    def spool_dir
      dir = ENV["REPROIT_CI_SPOOL"].to_s
      dir.empty? ? DEFAULT_SPOOL_DIR : dir
    end

    def spool_max_bytes
      parsed = Integer(ENV["REPROIT_CI_SPOOL_MAX"].to_s, 10, exception: false)
      return DEFAULT_SPOOL_MAX_BYTES if parsed.nil?
      parsed.clamp(SPOOL_MAX_FLOOR_BYTES, SPOOL_MAX_CEIL_BYTES)
    end

    def record_drop(dir)
      counter = File.join(dir, "dropped.count")
      dropped = begin
        Integer(File.read(counter).strip, 10, exception: false) || 0
      rescue SystemCallError
        # First drop: the counter does not exist yet.
        0
      end
      File.write(counter, (dropped + 1).to_s + "\n")
    end

    # Write one capsule inside the byte cap; over-cap capsules are dropped
    # and counted. Returns the file path or nil.
    def spool(payload)
      body = ReproitBackendRb.canonical_json(payload)
      dir = spool_dir
      FileUtils.mkdir_p(dir)
      used = Dir.children(dir).sum do |entry|
        next 0 unless entry.end_with?(".json")
        begin
          File.size(File.join(dir, entry))
        rescue SystemCallError
          # A concurrently removed entry counts as zero.
          0
        end
      end
      if used + body.bytesize > spool_max_bytes
        count(:dropped_capsules)
        record_drop(dir)
        return nil
      end
      digest = Digest::SHA256.hexdigest(body)[0, 12]
      file = File.join(dir, "capsule-" + digest + ".json")
      File.write(file, body)
      count(:spooled_capsules)
      $stderr.write(
        SPOOL_MARKER + JSON.generate({ "file" => file, "operation" => payload["operation"] }) +
        "\n"
      )
      file
    end

    def finish_and_spool(trace, operation, error)
      trace.finish({ "error" => bounded_error(error) }, nil, false, false)
      spool({
        "format" => CAPTURE_FORMAT,
        "version" => CAPTURE_VERSION_EXCHANGES,
        "operation" => operation,
        "oracle" => TEST_FAILURE_ORACLE,
        "envelope" => envelope_for(trace),
        "events" => trace.events,
      })
    rescue StandardError
      # Capture must never mask the test's own failure.
      count(:failed_captures)
    end

    # The capsule names exactly one test; everything else is skipped so the
    # process exit code speaks for the named test alone.
    def replay_target
      @lock.synchronize do
        unless @replay_target_loaded
          @replay_target_loaded = true
          payload = JSON.parse(File.read(replay_path))
          operation = payload["operation"]
          unless operation.is_a?(String) && operation.start_with?(TEST_TRIGGER_PREFIX)
            raise ArgumentError,
                  "REPROIT_REPLAY capsule does not carry a test trigger identity"
          end
          @replay_target = operation
        end
        @replay_target
      end
    end

    def report_result(operation, status, error)
      detail = { "operation" => operation, "status" => status }
      detail["failure"] = bounded_error(error) unless error.nil?
      $stderr.write(RESULT_MARKER + JSON.generate(detail) + "\n")
    end

    # Every test runs inside its own trace with the Net::HTTP hook and db
    # helper live, so dependency exchanges and the envelope record exactly as
    # production capture does; a failing test spools the capsule and
    # re-raises, never masking the failure.
    def capture_body(operation, suite_name, test_name, fn)
      proc do
        trace = BackendTrace.begin(
          CI.ci_context, operation,
          input: { "suite" => CI.bounded_name(suite_name), "test" => CI.bounded_name(test_name) }
        )
        begin
          Instrument.with_trace(trace) { instance_exec(&fn) }
        rescue Minitest::Skip
          # A skipped test asserted nothing; there is no invariant to spool.
          raise
        rescue Exception => error
          # Minitest::Assertion descends from Exception, not StandardError,
          # so a bare rescue would miss the very failures this mode exists
          # to capture. The error is always re-raised.
          CI.finish_and_spool(trace, operation, error)
          raise
        end
        begin
          trace.finish(nil, nil, true, false)
        rescue TraceError
          # An over-long passing trace has nothing to spool anyway.
        end
      end
    end

    def replay_body(operation, target, fn)
      return proc { skip("reproit replay targets " + target) } if operation != target
      proc do
        begin
          instance_exec(&fn)
        rescue Minitest::Skip
          raise
        rescue Exception => error
          CI.report_result(operation, "failed", error)
          raise
        end
        CI.report_result(operation, "passed", nil)
      end
    end

    # One Minitest::Test subclass per suite, running tests in DECLARATION
    # order (never shuffled): the flaky wedge exists to capture
    # order-dependent state, so the capture run and the replay run must walk
    # the tests the same way, like the Node reference's node:test does.
    def suite_class(suite_name)
      seq = @lock.synchronize { @suite_seq += 1 }
      klass = Class.new(Minitest::Test) do
        @reproit_methods = []

        def self.runnable_methods
          @reproit_methods.dup
        end

        def self.reproit_register(method_name, body)
          define_method(method_name, &body)
          @reproit_methods << method_name
        end
      end
      # A named constant so Minitest reports read `ReproitCiSuiteN#test_...`
      # instead of an anonymous class.
      const_set("ReproitCiSuite#{seq}", klass)
      klass
    end

    # `options` is reserved; there are none yet and unknown keys are
    # rejected so a typo cannot silently change capture behavior.
    def suite(suite_name, options = {})
      unless options.empty?
        raise ArgumentError, "reproit ci.suite: unknown option #{options.keys.first}"
      end
      active = mode
      Instrument.install unless active == "off"
      target = active == "replay" ? replay_target : nil
      klass = suite_class(suite_name)
      lambda do |test_name, &fn|
        operation = CI.operation_for(suite_name, test_name)
        method_name = format(
          "test_%03d_%s",
          klass.runnable_methods.length,
          test_name.to_s.gsub(/[^A-Za-z0-9]+/, "_")
        )
        body =
          case active
          when "capture" then CI.capture_body(operation, suite_name, test_name, fn)
          when "replay" then CI.replay_body(operation, target, fn)
          else fn
          end
        klass.reproit_register(method_name, body)
        method_name
      end
    end
  end
end

File: sdk/reproit-backend-rs/src/ci.rs
//! CI capture mode (feature `instrument`): the flaky-CI wedge.
//!
//! Rust port of the Node SDK's `ci.js`. Wrap a test body in
//! [`run`]`(suite, test, body)` and the trigger identity is the TEST (suite
//! plus test id), not an inbound HTTP request. With `REPROIT_CI_CAPTURE=1`
//! the body runs inside its own trace, so the instrumented outbound
//! boundaries (`instrument::http`, `db::run`, `pg`) record dependency
//! exchanges and the determinism envelope exactly as production capture
//! does; a FAILING test spools a version-2 `reproit-backend-capture` capsule
//! to a bounded on-disk spool. With `REPROIT_REPLAY` set the SAME wrapper
//! re-runs only the capsule's named test while the SDK serves the recorded
//! exchanges in process, and reports the observed result as a structured
//! stderr marker for `reproit check`. Without either env the wrapper runs
//! the body untouched.
//!
//! The wire is the existing capture payload: the test identity rides in the
//! `operation` field as `test:<suite>#<test>`, and the failed assertion is
//! the existing `backend-authored-invariant` registry oracle (a test IS an
//! authored invariant). No new protocol fields, no new oracle ids.
//!
//! Named deviations from the Node reference, forced by cargo test's model:
//!
//! - Adoption is PER TEST BODY, not a runner-level `test()` replacement: a
//!   Rust library cannot intercept `#[test]` functions, so each test wraps
//!   its body in `ci::run` the way an app adopts a logger. An unwrapped test
//!   is invisible to capture, exactly like an unwrapped client.
//! - libtest runs tests in parallel by default; order-dependent suites need
//!   `cargo test -- --test-threads=1` (sequential, name-sorted), which the
//!   SDK cannot impose from library code.
//! - Failure identity is the panic payload (`assert!` messages). Tests that
//!   fail by aborting the process, or `#[should_panic]` inversions, are not
//!   capturable at this layer.
//! - In replay mode non-target tests are skipped silently (their bodies do
//!   not run and libtest reports them as passed); node:test marks them
//!   skipped. The process exit code still speaks for the named test alone.
//!
//! Markers are written straight to the stderr handle so libtest's output
//! capture cannot swallow them (`--nocapture` is not required).
//!
//! Honest limit: replay pins the envelope and the recorded exchanges, which
//! is the whole boundary this SDK can see. A race the boundary cannot see
//! (scheduling, shared memory) is not reproduced by this capsule; `reproit
//! check` reports such runs Inconclusive, never a fake reproduction.

use crate::capture::{determinism_envelope, valid_token, CAPTURE_FORMAT};
use crate::framework::Recorder;
use crate::{instrument, BackendTrace, TraceContext};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::future::Future;
use std::io::Write;
use std::path::PathBuf;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::OnceLock;

/// Test-trigger identity prefix inside the existing `operation` field.
pub const TEST_TRIGGER_PREFIX: &str = "test:";
/// The registry oracle a failed test capsule carries: an authored invariant
/// (the test's own assertion) was violated. Existing id, not a new one.
pub const TEST_FAILURE_ORACLE: &str = "backend-authored-invariant";
/// Structured stderr markers `reproit check` parses, like REPROIT:DIVERGENCE.
pub const RESULT_MARKER: &str = "REPROIT:CI-TEST ";
pub const SPOOL_MARKER: &str = "REPROIT:CI-CAPSULE ";

/// Spool bounds. The cap covers the TOTAL bytes on disk; capsules beyond it
/// are dropped and counted (in-process stats plus the on-disk
/// `dropped.count`), never silently.
pub const DEFAULT_SPOOL_DIR: &str = ".reproit/ci-spool";
pub const DEFAULT_SPOOL_MAX_BYTES: u64 = 16 * 1024 * 1024;
const SPOOL_MAX_FLOOR_BYTES: u64 = 4 * 1024;
const SPOOL_MAX_CEIL_BYTES: u64 = 64 * 1024 * 1024;
/// Suite and test names share the operation field's 256-code-point bound.
const MAX_NAME: usize = 120;
const MAX_ERROR_CHARS: usize = 2048;

static TRACE_SEQ: AtomicU64 = AtomicU64::new(1);
static SPOOLED: AtomicU64 = AtomicU64::new(0);
static DROPPED: AtomicU64 = AtomicU64::new(0);
static FAILED_CAPTURES: AtomicU64 = AtomicU64::new(0);

/// CI capture counters, the Node reference's `ci.stats()`.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct CiStats {
    pub spooled_capsules: u64,
    pub dropped_capsules: u64,
    pub failed_captures: u64,
}

pub fn stats() -> CiStats {
    CiStats {
        spooled_capsules: SPOOLED.load(Ordering::Relaxed),
        dropped_capsules: DROPPED.load(Ordering::Relaxed),
        failed_captures: FAILED_CAPTURES.load(Ordering::Relaxed),
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Mode {
    Off,
    Capture,
    Replay,
}

fn replay_path() -> Option<String> {
    std::env::var("REPROIT_REPLAY")
        .ok()
        .filter(|path| !path.trim().is_empty())
}

fn mode() -> Mode {
    if replay_path().is_some() {
        return Mode::Replay;
    }
    if std::env::var("REPROIT_CI_CAPTURE").ok().as_deref() == Some("1") {
        return Mode::Capture;
    }
    Mode::Off
}

fn bounded_name(value: &str) -> String {
    value.trim().chars().take(MAX_NAME).collect()
}

/// The capsule's operation identity: `test:<suite>#<test>`.
pub fn operation_for(suite: &str, test: &str) -> String {
    format!(
        "{TEST_TRIGGER_PREFIX}{}#{}",
        bounded_name(suite),
        bounded_name(test)
    )
}

fn bounded_error(message: &str) -> String {
    message.chars().take(MAX_ERROR_CHARS).collect()
}

/// The failure identity of a panicked test body: its panic payload.
fn panic_message(payload: Box<dyn std::any::Any + Send>) -> String {
    if let Some(text) = payload.downcast_ref::<String>() {
        return text.clone();
    }
    if let Some(text) = payload.downcast_ref::<&str>() {
        return (*text).to_string();
    }
    "test panicked".to_string()
}

/// Synthesized trace context: the CI job stands where production stood.
fn ci_context() -> TraceContext {
    let build = ["REPROIT_COMMIT", "GITHUB_SHA"]
        .iter()
        .find_map(|name| std::env::var(name).ok().filter(|value| valid_token(value)));
    TraceContext {
        trace_id: format!(
            "ci-{}-{}",
            crate::capture::now_millis(),
            TRACE_SEQ.fetch_add(1, Ordering::Relaxed)
        ),
        actor: None,
        action_index: 0,
        build,
        config_contract: None,
        capture_envelope: true,
        replay_seed: Some(format!("{:016x}", crate::capture::now_millis() | 1)),
    }
}

/// Straight to fd 2: libtest's output capture reroutes the print macros to a
/// buffer it re-prints on STDOUT, where `reproit check` (watching stderr)
/// could never see a marker.
fn stderr_line(line: &str) {
    let mut stderr = std::io::stderr().lock();
    let _ = writeln!(stderr, "{line}");
}

fn spool_dir() -> PathBuf {
    match std::env::var("REPROIT_CI_SPOOL") {
        Ok(dir) if !dir.is_empty() => PathBuf::from(dir),
        _ => PathBuf::from(DEFAULT_SPOOL_DIR),
    }
}

fn spool_max_bytes() -> u64 {
    match std::env::var("REPROIT_CI_SPOOL_MAX")
        .ok()
        .and_then(|value| value.trim().parse::<u64>().ok())
    {
        Some(parsed) => parsed.clamp(SPOOL_MAX_FLOOR_BYTES, SPOOL_MAX_CEIL_BYTES),
        None => DEFAULT_SPOOL_MAX_BYTES,
    }
}

fn record_drop(dir: &std::path::Path) -> std::io::Result<()> {
    let counter = dir.join("dropped.count");
    // First drop: the counter does not exist yet.
    let dropped = std::fs::read_to_string(&counter)
        .ok()
        .and_then(|text| text.trim().parse::<u64>().ok())
        .unwrap_or(0);
    std::fs::write(&counter, format!("{}\n", dropped + 1))
}

/// Write one capsule inside the byte cap; over-cap capsules are dropped and
/// counted. serde_json's BTreeMap order makes the bytes canonical, matching
/// the Node reference's canonicalJson.
fn spool(payload: &Value, operation: &str) -> std::io::Result<Option<PathBuf>> {
    let body = serde_json::to_vec(payload).map_err(std::io::Error::other)?;
    let dir = spool_dir();
    std::fs::create_dir_all(&dir)?;
    let mut used = 0u64;
    for entry in std::fs::read_dir(&dir)? {
        let entry = entry?;
        if entry.path().extension().and_then(|ext| ext.to_str()) != Some("json") {
            continue;
        }
        // A concurrently removed entry counts as zero.
        used += entry.metadata().map(|meta| meta.len()).unwrap_or(0);
    }
    if used + body.len() as u64 > spool_max_bytes() {
        DROPPED.fetch_add(1, Ordering::Relaxed);
        record_drop(&dir)?;
        return Ok(None);
    }
    let digest = Sha256::digest(&body);
    let short: String = digest[..6]
        .iter()
        .map(|byte| format!("{byte:02x}"))
        .collect();
    let file = dir.join(format!("capsule-{short}.json"));
    std::fs::write(&file, &body)?;
    SPOOLED.fetch_add(1, Ordering::Relaxed);
    // Field order mirrors the Node reference's insertion order.
    stderr_line(&format!(
        "{SPOOL_MARKER}{{\"file\":{},\"operation\":{}}}",
        json!(file.display().to_string()),
        json!(operation),
    ));
    Ok(Some(file))
}

fn finish_and_spool(recorder: Recorder, operation: &str, message: &str) {
    // Capture must never mask the test's own failure: every fallible step
    // lands in failed_captures instead of propagating.
    let Some(mut trace) = recorder.into_trace() else {
        FAILED_CAPTURES.fetch_add(1, Ordering::Relaxed);
        return;
    };
    if trace
        .finish_test(json!({ "error": message }), false)
        .is_err()
    {
        FAILED_CAPTURES.fetch_add(1, Ordering::Relaxed);
        return;
    }
    // Same envelope shape production capture records; the seed pins the
    // REPLAY run's randomness, it does not reproduce the test run's.
    let observed_at = trace
        .events()
        .first()
        .and_then(|event| event.get("at"))
        .and_then(Value::as_u64);
    let payload = json!({
        "format": CAPTURE_FORMAT,
        "version": 2,
        "operation": operation,
        "oracle": TEST_FAILURE_ORACLE,
        "envelope": determinism_envelope(observed_at),
        "events": trace.events(),
    });
    if spool(&payload, operation).is_err() {
        FAILED_CAPTURES.fetch_add(1, Ordering::Relaxed);
    }
}

/// The one test the loaded capsule names; a capsule without a test trigger
/// identity fails closed (every wrapped test panics with the reason).
fn replay_target() -> &'static str {
    static TARGET: OnceLock<String> = OnceLock::new();
    TARGET.get_or_init(|| {
        let path = replay_path().expect("replay mode without REPROIT_REPLAY");
        let text = std::fs::read_to_string(&path)
            .unwrap_or_else(|error| panic!("reproit ci replay refused: read {path}: {error}"));
        let payload: Value = serde_json::from_str(&text)
            .unwrap_or_else(|error| panic!("reproit ci replay refused: not JSON: {error}"));
        let operation = payload
            .get("operation")
            .and_then(Value::as_str)
            .unwrap_or_default();
        assert!(
            operation.starts_with(TEST_TRIGGER_PREFIX),
            "REPROIT_REPLAY capsule does not carry a test trigger identity"
        );
        operation.to_string()
    })
}

fn report_result(operation: &str, status: &str, failure: Option<&str>) {
    // Hand-assembled so the field order matches the Node reference's
    // insertion order byte for byte.
    let mut line = format!(
        "{RESULT_MARKER}{{\"operation\":{},\"status\":{}",
        json!(operation),
        json!(status),
    );
    if let Some(failure) = failure {
        line.push_str(&format!(",\"failure\":{}", json!(failure)));
    }
    line.push('}');
    stderr_line(&line);
}

/// Run one test body under CI capture/replay, the Node reference's
/// `ci.suite(name)(test, fn)`. Call it as the whole body of a `#[tokio::test]`
/// (or any test running inside a tokio runtime):
///
/// ```ignore
/// #[tokio::test]
/// async fn order_total_applies_the_tax_rate() {
///     reproit_backend::ci::run("checkout", "order total applies the tax rate", async {
///         assert_eq!(order_total(100.0).await, 125.0);
///     })
///     .await;
/// }
/// ```
///
/// Off mode runs the body untouched. Capture mode scopes it over a fresh
/// trace (the instrument boundaries record exchanges) and spools a capsule
/// when it panics, then re-raises the panic so the test still fails. Replay
/// mode runs ONLY the capsule's named test and reports the observed result
/// as the `REPROIT:CI-TEST` stderr marker.
pub async fn run<F>(suite: &str, test: &str, body: F)
where
    F: Future<Output = ()> + Send + 'static,
{
    match mode() {
        Mode::Off => body.await,
        Mode::Capture => capture_run(suite, test, body).await,
        Mode::Replay => replay_run(suite, test, body).await,
    }
}

async fn capture_run<F>(suite: &str, test: &str, body: F)
where
    F: Future<Output = ()> + Send + 'static,
{
    let operation = operation_for(suite, test);
    let begun = BackendTrace::begin(
        ci_context(),
        operation.as_str(),
        None,
        None,
        None,
        json!({ "suite": bounded_name(suite), "test": bounded_name(test) }),
        Vec::new(),
    );
    let Ok(trace) = begun else {
        // Capture failed to start; the test itself must still run and speak.
        FAILED_CAPTURES.fetch_add(1, Ordering::Relaxed);
        body.await;
        return;
    };
    let recorder = Recorder::standalone(trace);
    // tokio::spawn is the panic boundary: a panicking body surfaces as a
    // JoinError instead of unwinding through this frame, so the capsule
    // spools BEFORE the failure re-raises.
    let outcome = tokio::spawn(instrument::scope(recorder.clone(), body)).await;
    match outcome {
        Ok(()) => {
            // An over-long passing trace has nothing to spool anyway.
            if let Some(mut trace) = recorder.into_trace() {
                let _ = trace.finish_test(Value::Null, true);
            }
        }
        Err(joined) => {
            let message = bounded_error(&if joined.is_panic() {
                panic_message(joined.into_panic())
            } else {
                "test aborted".to_string()
            });
            finish_and_spool(recorder, &operation, &message);
            std::panic::resume_unwind(Box::new(message));
        }
    }
}

async fn replay_run<F>(suite: &str, test: &str, body: F)
where
    F: Future<Output = ()> + Send + 'static,
{
    // Load the session now so the envelope (TZ, clock offset, seed) is
    // pinned before any test code runs.
    instrument::init();
    let operation = operation_for(suite, test);
    if operation != replay_target() {
        // The capsule names exactly one test; everything else is skipped so
        // the process exit code speaks for the named test alone.
        return;
    }
    let outcome = tokio::spawn(body).await;
    match outcome {
        Ok(()) => report_result(&operation, "passed", None),
        Err(joined) => {
            let message = bounded_error(&if joined.is_panic() {
                panic_message(joined.into_panic())
            } else {
                "test aborted".to_string()
            });
            report_result(&operation, "failed", Some(&message));
            std::panic::resume_unwind(Box::new(message));
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn operation_identity_is_bounded_and_prefixed() {
        assert_eq!(
            operation_for(" checkout ", "order total"),
            "test:checkout#order total"
        );
        let long = "x".repeat(400);
        let operation = operation_for(&long, &long);
        assert_eq!(
            operation.chars().count(),
            TEST_TRIGGER_PREFIX.len() + MAX_NAME * 2 + 1
        );
    }

    #[test]
    fn result_marker_matches_the_node_field_order() {
        // Byte parity with node's `JSON.stringify({operation, status, failure})`.
        let mut line = format!(
            "{RESULT_MARKER}{{\"operation\":{},\"status\":{}",
            json!("test:s#t"),
            json!("failed"),
        );
        line.push_str(&format!(",\"failure\":{}", json!("7 != 8")));
        line.push('}');
        assert_eq!(
            line,
            "REPROIT:CI-TEST {\"operation\":\"test:s#t\",\"status\":\"failed\",\
             \"failure\":\"7 != 8\"}"
        );
    }

    #[test]
    fn panic_payloads_and_bounds_shape_the_failure_identity() {
        assert_eq!(panic_message(Box::new("boom")), "boom");
        assert_eq!(panic_message(Box::new("boom".to_string())), "boom");
        assert_eq!(panic_message(Box::new(7u8)), "test panicked");
        assert_eq!(
            bounded_error(&"e".repeat(4096)).chars().count(),
            MAX_ERROR_CHARS
        );
    }
}
